File: src/data/types/sellToMarketResult.ts
import type { Wallet } from "./wallet.ts";
import type { PlanetMarketState } from "./planetMarketState.ts";

// Agent 11 amendment (Trading Counterparty). Not a discriminated union like
// PurchaseResult -- sellToMarket() throws for its one failure mode
// (non-positive quantity) rather than returning a typed rejection, since
// selling your own inventory has no normal business-rejection case the way
// purchaseListing()'s self-trade/over-quantity checks do (those react to
// another player's independent action). Every successful call returns this
// shape directly, same as createListing() returning Listing directly.
export interface SellToMarketResult {
  quantitySold: number;
  totalValue: number;
  feeDeducted: number;
  proceedsToSeller: number;
  updatedWallet: Wallet;
  updatedMarketState: PlanetMarketState;
}

File: unity/ProfitableCore/Schema/SellToMarketResult.cs
namespace Profitable.Core.Schema;

// Ports src/data/types/sellToMarketResult.ts. Not a discriminated union
// like PurchaseResult -- SellToMarket throws for its one failure mode
// (non-positive quantity) rather than returning a typed rejection, since
// selling your own inventory has no normal business-rejection case the
// way PurchaseListing's self-trade/over-quantity checks do (those react
// to another player's independent action).
public class SellToMarketResult
{
    public int QuantitySold { get; init; }
    public double TotalValue { get; init; }
    public double FeeDeducted { get; init; }
    public double ProceedsToSeller { get; init; }
    public Wallet UpdatedWallet { get; init; } = new();
    public PlanetMarketState UpdatedMarketState { get; init; } = new();
}
